rfcs/web_extensions.md

```diff
@@ -0,0 +1,137 @@
+a web extension designed to test the functionality of a specific API.
+
+These tests won't leverage `testharness.js` directly since most of the test execution is handled
+within the extension, via the `browser.test` API. We’ve elected to use this API since all participating
+browser vendors use `browser.test` internally and can easily port over existing tests to the
+web-platform-tests.
+
+The `browser.test` API will also be used to send messages back to the web page to notify the
+testharness when a test has started and finished. For these cases, it is up to the user agent to
+send the message "test-started" and "test-finished", respectively. In addition, when recording the
+result for  a `browser.test.assert...()` method, it is up to the user agent to send the message
+"assert" or "assert-equality".
+
+Because these tests don’t leverage `testharness.js` directly, we’ve introduced a new wrapper type,
+`.extension.js`, that will create the necessary boilerplate to convert these messages into the
+corresponding assertions in the `testharness.js`.
+
+It's important to note that it isn't guaranteed that the web page will load before the extension loads.
+Meaning, a race condition could occur that would cause the `browser.test` assertion results to not
+be reported back to the test harness, since the `browser.test.onMessage` listener wasn't registered.
+To resolve this, user agents would need to queue the results until a listener is registered, and then
+report them.
+
+Last, Since the `Classic` and `BiDi` implementations support loading an extension using a `path` or
+`archivePath` to the extension's resources, the full path to the extension's resources is formulated
+by using the `test.path` before the protocol loads the extension.
+
+A proposed patch is available at https://github.com/web-platform-tests/wpt/pull/50648.
+
+##  `browser.test` API
+
+The `browser.test` API provides utilities for writing and running tests for WebExtensions.
+It includes assertion functions and a test harness to facilitate structured and reliable testing.
+
+### Methods
+
+**`browser.test.assertTrue(condition, message)`**\
+Asserts that a given condition is `true`. If the condition is `false`,
+the test fails.
+
+**Parameters**
+
+* `condition` (boolean) – The condition to assert. If `false`, the test fails.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.assertFalse(condition, message)`**\
+Asserts that a given condition is `false`. If the condition is `true`, the test fails.
+
+**Parameters**
+
+* `condition` (boolean) – The condition to assert. If `true`, the test fails.
+* `message` (string, optional) – A message describing the assertion.
+
+\
+**`browser.test.assertEq(actual, expected, message)`**\
+Asserts that the actual value is equal to the expected value. If not equal, the test fails.\
+This matches the ES [SameValue](https://tc39.es/ecma262/#sec-samevalue) implementation.
```

Comment:
This is not true in Chromium-based browsers.  In Chromium-based browsers, we have custom handling for anything with `typeof` equal to `"object"`, which then results in a "deep equals" comparison (code [here](https://source.chromium.org/chromium/chromium/src/+/main:extensions/renderer/resources/test_custom_bindings.js;l=248-252;drc=abe36fd704986c2e4f54b4bd35c291d6beb0e28c)).  This is important, since it enables things like:

```
browser.test.assertEq({foo: 3, bar: 2}, value);
```

Reply:
Both we and Safari have a separate `test.assertDeepEq()` for this use case.

I don't have a strong reason to oppose changing the `assertEq()` behavior to match Chromium, but there's something to be said for less magic and more explicitness in API design.